http/src/v0/root_files/tar_helper.rs

```diff
@@ -0,0 +1,332 @@
+                return Err(GetError::InvalidLinkName(data.to_vec()));
+            }
+
+            // this is another long header trick, but this time we have different entry type but
+            // similarly the long file name is written as an separate entry with own headers.
+
+            self.long_filename_header.set_size(data.len() as u64 + 1);
+            self.long_filename_header
+                .set_entry_type(tar::EntryType::new(b'K'));
+            self.long_filename_header.set_cksum();
+
+            self.bytes.put_slice(self.long_filename_header.as_bytes());
+            self.bytes.put_slice(data);
+            self.bytes.put_u8(0);
+            ret[2] = Some(self.bytes.split().freeze());
+
+            ret[3] = self.pad(data.len() as u64 + 1);
+        }
+
+        Self::set_metadata(&mut self.header, metadata, 0o0644);
+        self.header.set_size(0);
+        self.header.set_entry_type(tar::EntryType::Symlink);
+        self.header.set_cksum();
+
+        self.bytes.put_slice(self.header.as_bytes());
+        ret[4] = Some(self.bytes.split().freeze());
+
+        Ok(ret)
+    }
+
+    /// Content is tar files is padded to 512 byte sectors which might be configurable as well.
+    pub(super) fn pad(&self, total_size: u64) -> Option<Bytes> {
+        let padding = 512 - (total_size % 512);
+        if padding < 512 {
+            Some(self.zeroes.slice(..padding as usize))
+        } else {
+            None
+        }
+    }
+
+    fn set_metadata(header: &mut tar::Header, metadata: &Metadata, default_mode: u32) {
+        header.set_mode(
+            metadata
+                .mode()
+                .map(|mode| mode & 0o7777)
+                .unwrap_or(default_mode),
+        );
+
+        header.set_mtime(
+            metadata
+                .mtime()
+                .and_then(|(seconds, _)| {
+                    if seconds >= 0 {
+                        Some(seconds as u64)
+                    } else {
+                        None
+                    }
+                })
+                .unwrap_or(0),
+        );
+    }
+}
+
+/// Returns the raw bytes we need to write as a new entry into the tar.
+fn prepare_long_header<'a>(
+    header: &mut tar::Header,
+    long_filename_header: &mut tar::Header,
+    path: &'a Path,
+    _error: std::io::Error,
+) -> Result<&'a [u8], GetError> {
+    #[cfg(unix)]
+    /// On unix this operation can never fail.
+    pub(super) fn bytes2path(bytes: Cow<[u8]>) -> std::io::Result<Cow<Path>> {
+        use std::ffi::{OsStr, OsString};
+        use std::os::unix::prelude::*;
+
+        Ok(match bytes {
+            Cow::Borrowed(bytes) => Cow::Borrowed(Path::new(OsStr::from_bytes(bytes))),
+            Cow::Owned(bytes) => Cow::Owned(PathBuf::from(OsString::from_vec(bytes))),
+        })
+    }
+
+    #[cfg(windows)]
+    /// On windows we cannot accept non-Unicode bytes because it
+    /// is impossible to convert it to UTF-16.
```

Comment:
```suggestion
    /// is impossible to convert them to UTF-16.
```

Reply:
Ignoring this to keep as much 1:1 with upstream where this was copied from: https://github.com/alexcrichton/tar-rs/blob/0b18e421f3846e887111e4a8e964f7291367ada6/src/header.rs#L1566-L1587